### ISSUE TITLE
【模型-压缩】模型压缩方法概述

### ISSUE THREAD

**Issue by QingyaFan** [QingyaFan/blog#217]
# 【模型-压缩】模型压缩方法概述

## 为什么要压缩模型

压缩模型是因为现在的模型太"胖"了，而硬件（特别是显存和钱包）又太"瘦"了。如果把训练出一个强大的大模型比做"修建一座图书馆"，那么模型压缩就是把这座图书馆塞进一个"Kindle"里。压缩模型的具体动机主要有：

1. 突破显存瓶颈（Fitting the Elephant）：主流大模型通常以 FP16 精度存储，如果不压缩，70B的模型就需要约 140G 的显存，而A100也就单卡80G显存。
2. 提升推理速度（Speed is Everything）：模型越小，计算量和内存带宽的需求就越低。
3. 降低部署成本（ROI）：在云端部署 AI 服务极其昂贵。
4. 走向端侧设备（Edge Computing）

模型压缩本质上是精度（Accuracy）、速度（Latency）和资源（Resources）之间寻找最优平衡点。没有压缩技术，大模型就是实验室里的昂贵玩具，有了压缩，它才能成为每个人手里随处可见的工具。

### 不同规模模型的计算性能差异

以 Qwen3-Embedding-4B 和 Qwen3-Embedding-8B 为例，两者同属 Qwen3 架构，具体参数对比：

| 参数 | Qwen3-Embedding-4B | Qwen3-Embedding-8B |
|--|--|--|
| 层数（num_hidden_layers） | 36 | 36 |
| 隐藏维度（hidden_size） | 2560 | 4096 |
| 注意力头数（num_attention_heads） | 32 | 32 |
| KV 头数（num_key_value_heads） | 8 | 8 |
| FFN 中间维度（intermediate_size） | 9728 | 14336 |
| 输出 Embedding 维度 | 2560 | 4096 |

可以看到两者层数相同（都是 36 层），差异主要在 **hidden_size**（2560 vs 4096）和 **intermediate_size**（9728 vs 14336）。这意味着每一层中的矩阵乘法规模 8B 显著大于 4B。

**计算性能影响：**

- **推理吞吐量**：每层的计算量与 hidden_size² 和 intermediate_size × hidden_size 成正比。8B 的 hidden_size 是 4B 的 1.6 倍，因此单层 Self-Attention 的 QKV 投影计算量约为 4B 的 2.56 倍（1.6²）；FFN 层的计算量约为 4B 的 2.3 倍（14336×4096 / 9728×2560）。整体单次前向 FLOPs 约为 4B 的 2 倍。
- **单请求延迟**：层数相同，差异完全来自每层更宽的矩阵运算。对 Embedding 模型只有 encode pass（无自回归生成），延迟差异近似与每层计算量成正比。
- **显存带宽**：同一张卡上 4B 模型显存占用更低，可以跑更大的 batch size，进一步提升吞吐。

实际加速比通常在 **1.5x ~ 1.8x**（而非理论的 2x），因为部分开销是固定的（数据加载、tokenize、post-processing），且小 batch 下是 memory-bandwidth bound，大 batch 下才接近 compute bound。

**结论：选 4B 还是 8B 本质是质量 vs 速度/成本的权衡。** 如果 4B 在目标检索场景下精度够用（可通过 MTEB 等 benchmark 验证），那它在显存和计算性能上都是更优选择。

## 模型压缩方法

主要的模型压缩方法可以归纳为量化（Quantization）、剪枝（Pruning）、知识蒸馏（Knowledge Distillation）和低秩分解（Low-Rank Factorization）四大类。

前三种压缩方法综合对比总结：

|维度|量化(Quantization)|剪枝(Pruning)|蒸馏(Distillation)|
|--|--|--|--|
| 显存降低 | 极高（2x - 8x）| 中等（由剪枝率决定） | 极高（由架构决定）|
| 精度保持 | 较好（4-bit 以上几乎无损）| 视剪枝比例而定 | 优秀（取决于教师模型）|
| 计算开销 | 极低（PTQ 仅需几分钟 | 中到高（通常需要微调）| 极高（需要完整训练过程）|
| 主要目标 | 部署加速、显存优化 | 减少参数、架构瘦身 | 跨架构能力迁移 |

**通过对比可以看到，剪枝和蒸馏在 ROI 上低于量化，这也是这两项技术逐渐不太常用的原因。**

### 模型量化

量化是通过降低模型参数的数值精度（如从 FP32 降至 INT4 或 FP8）来减少显存占用和加速推理的技术。

核心方法与适用场景:

| 方法名称 | 技术特点 | 最佳适用场景 |
| -- | -- | -- |
|GGUF（GGML）|基于块的量化，支持 CPU/GPU 混合推理，元数据兼容性强。|个人电脑、Mac、边缘计算设备，尤其是显存不足以装下全模型的场景。|
|GPTQ|一次性（One-shot）权重向量量化，利用二阶导数信息最小化误差。|NVIDIA GPU生产环境，追求极致的吞吐量（Throughput）和推理速度。|
|AWQ| 激活感知权重量化，通过保护 1% 的显著权重（Salient Weights）维持精度。|创作类、代码类模型，对输出质量敏感且使用 NVIDIA/AMD GPU的场景。|
|bitsandbytes（NF4）|归一化浮点（NormalFloat4），利用权重分布规律进行的 4-bit 量化。|模型微调（QLoRA）及快速原型部署，无需离线校准数据集。|
|BitNet（1.58b）|极低比特量化（1-bit），参数仅含 `{-1, 0, 1}`|下一代端侧模型，追求理论上的极限能效比（Energy Efficiency）。|

#### 如何识别模型是否经过量化

从 Hugging Face 下载或引用一个模型时，可以通过以下五个维度判断它是否经过量化：

**1. `config.json` 中的 `quantization_config` 字段**

这是最可靠的判断依据。量化模型的 `config.json` 通常包含：

```json
"quantization_config": {
  "quant_method": "gptq",
  "bits": 4,
  "group_size": 128,
  "desc_act": true
}
```

常见的 `quant_method` 值包括 `gptq`、`awq`、`bitsandbytes`、`exl2` 等，`bits` 字段标明量化位宽。

需要注意：`quantization_config` 是 Hugging Face Transformers 库的序列化约定，而非跨生态通用协议。量化工具（AutoGPTQ、AutoAWQ、bitsandbytes 等）依赖 Transformers 的 `save_pretrained()` 保存模型时会自动写入此字段。不依赖 Transformers 的工具（如 llama.cpp/GGUF、TensorRT-LLM）有各自独立的元数据方案，不会产出这个字段。因此不能仅凭此单一维度判断，需结合下述多个信号交叉验证。

**2. 独立的 `quantize_config.json` 文件**

GPTQ 量化的模型通常会额外携带此文件，记录量化参数（bits、group_size、damp_percent 等）。看到这个文件基本可以确认是 GPTQ 量化产物。

**3. 文件名/模型名命名规范**

社区有约定俗成的命名标识：

| 标识示例 | 含义 |
|--|--|
| `GPTQ`、`AWQ`、`EXL2` | 对应量化方法 |
| `Q4_K_M`、`Q5_0`、`Q8_0` | GGUF 格式的量化等级 |
| `int4`、`int8`、`4bit`、`8bit` | 位宽标识 |

例如：`Qwen2-7B-Instruct-GPTQ-Int4`、`Meta-Llama-3-8B-Q4_K_M.gguf`。

**4. 文件大小推算**

粗略换算：1B 参数在 FP16 下约占 2GB，INT4 下约占 0.5GB。如果一个 7B 模型的权重文件总大小约 4GB 而非 14GB，大概率经过了 4-bit 量化。

**5. Model Card（README.md）说明**

正规发布的量化模型会在 Model Card 中注明原始模型来源、量化方法和工具（如 AutoGPTQ、llama.cpp）、量化精度等信息。这也是快速筛选的入口。

#### 运行时精度：量化的是权重，不一定是计算

谈量化模型时容易混淆三个精度概念，需要分开看：

| 维度 | 含义 | 是否随量化方法变化 |
|--|--|--|
| 权重存储精度（weight dtype） | 权重在显存里以什么格式存 | 变（INT4/INT8/FP8/FP16/BF16） |
| 激活精度（activation dtype） | KV cache、hidden state、layernorm 输出等 | 多数情况仍是 FP16/BF16 |
| 计算精度（compute dtype） | 矩阵乘真正在 Tensor Core 上的执行精度 | 取决于 kernel |

vLLM 的 `--dtype` 参数控制的是激活精度，不是权重精度；权重精度由模型文件里的 `quantization_config` 决定。

常见量化方案在推理引擎中的精度组合：

| 方案 | 权重 | 激活 | 计算 | 备注 |
|--|--|--|--|--|
| 未量化（FP16） | FP16 | FP16 | FP16 | 全程 FP16 |
| GPTQ-Int4 / AWQ-Int4（W4A16） | INT4 | FP16 | FP16 | 在 GEMM kernel 内把 INT4 反量化成 FP16 后再算 |
| FP8（W8A8） | FP8 | FP8（动态量化） | FP8，累加到 FP32 | Hopper/Blackwell 原生支持 |
| INT8 SmoothQuant（W8A8） | INT8 | INT8 | INT8，累加到 INT32 | 完成后反量化回 FP16 |
| KV cache 量化 | 不影响权重 | KV 存为 FP8/INT8 | — | 与权重量化正交，需单独开关 |

由此可以得出几条实践结论：

- **主流的 W4A16（GPTQ/AWQ）只压权重不压激活**，所以激活和中间张量默认仍是 FP16/BF16。它节省的是显存容量和带宽，不是算力。
- **A100 上跑 GPTQ-Int4 不一定比 FP16 快**，原因就在这里——权重要先反量化成 FP16 才能进 Tensor Core，理论 FLOPs 不变，加速主要来自带宽降低，反量化 kernel 开销可能抵消收益。Marlin、Machete 等 kernel 的优化目标就是这一步。
- **真正想提速**，需要 FP8（H100+）或 INT8 W8A8，让 Tensor Core 直接吃低精度数据。
- **显存估算要分两块**：模型权重按量化位宽算，KV cache 按激活精度算（默认 FP16），别只看权重大小忽略 KV。

一句话记忆：**量化是"压权重不压激活"的简称，运行时计算大头仍是 FP16；只有 FP8、W8A8、KV 量化才真正改变运行时精度。**

#### 主流推理引擎的运行时精度策略对比

上面以 vLLM 为例讲了基本规律。其他主流引擎在权重支持、计算路径、调度方式上有各自的取舍，下面做一次横向对比。

| 引擎 | 主要量化格式 | 权重→计算路径 | 激活/KV 精度 | 编译方式 | 定位 |
|--|--|--|--|--|--|
| **vLLM** | GPTQ、AWQ、compressed-tensors（INT8/FP8 W8A8）、bitsandbytes、Marlin/Machete、AQLM、HQQ、GGUF（实验） | W4A16：kernel 内反量化到 FP16；W8A8/FP8：原生低精度 GEMM | 默认 FP16/BF16；KV 可选 FP8（E4M3/E5M2）、INT8 | JIT，PyTorch + 自定义 CUDA kernel | 通用云端推理事实标准 |
| **SGLang** | 与 vLLM 大体同源（GPTQ/AWQ/FP8/INT8 W8A8/compressed-tensors），共享 Marlin/Machete | 与 vLLM 一致：W4A16 反量化为 FP16，FP8 走原生 Tensor Core | FP16/BF16；KV 支持 FP8 | JIT | 强项在 RadixAttention 和调度，量化路径基本复用 vLLM 生态 |
| **TGI**（HF Text Generation Inference） | GPTQ、AWQ、bitsandbytes（NF4/INT8）、EETQ、Marlin、FP8（Hopper） | W4A16：反量化到 FP16；FP8：原生 FP8 GEMM | FP16/BF16 为主；KV 暂以 FP16/BF16 为主 | JIT | HF 官方服务化方案，与 Transformers 生态衔接最紧 |
| **TensorRT-LLM** | FP8（W8A8）、INT4 AWQ、SmoothQuant W8A8、INT8 KV，由 ModelOpt 工具链产出 | AOT 编译，将 GEMM、反量化、激活量化融合到 TRT engine 里 | 整张图按计划好的精度跑，FP8/INT8 路径可端到端贯通 | AOT 编译为 engine，不可热加载 | NVIDIA GPU 上的性能上限，工程门槛最高 |
| **llama.cpp / GGUF** | Q2_K~Q8_0、IQ2/IQ3/IQ4（importance matrix） | dequant 到 FP16 或 FP32 后做 GEMM；CPU 端累加多用 FP32 | CPU 上激活通常 FP32，GPU/Metal 上 FP16；KV 可量化为 Q4_0/Q8_0/F16 | JIT，CPU/GPU/Metal 多后端 | 端侧、Mac、CPU 私有部署的事实标准 |

几个值得注意的差异：

- **vLLM 与 SGLang 的精度策略基本同源**。SGLang 大量复用 compressed-tensors / Marlin 这类 kernel，所以同一份 GPTQ/AWQ/FP8 模型在两边的运行时精度几乎一致，主要差距在调度和 KV 管理。
- **TGI 与 vLLM 的差距在收敛**。早期 TGI 量化能力弱于 vLLM，近期版本通过引入 Marlin、FP8、compressed-tensors 已基本对齐 W4A16 / W8A8 路径。
- **TensorRT-LLM 是唯一真正做"端到端低精度"的引擎**。它在 AOT 阶段就把激活量化、反量化、GEMM、bias、激活函数融合成一个图，FP8/INT8 可以贯穿整层而不是 kernel 内进出 FP16，因此在 H100 上 FP8 吞吐通常显著高于 vLLM。代价是不可热加载，每次模型/序列长度配置变化都要重编译。
- **llama.cpp 是另一条技术路线**。它不依赖 Tensor Core，CPU 上靠 AVX/NEON + FP32 累加，GPU 上才会 dequant 到 FP16。所以谈"运行时精度"时，llama.cpp 的激活精度往往比云端引擎更高（FP32），但吞吐受限于反量化和访存。
- **KV cache 量化是各家差异最大的一块**。vLLM/SGLang 已较成熟支持 FP8 KV，TensorRT-LLM 支持 INT8 KV，TGI 较保守，llama.cpp 走 GGUF 自有的 KV 量化体系。在长上下文场景下，KV 精度选择对显存的影响往往大于权重量化。

选型上的简单经验：

- **追求吞吐上限、能接受工程复杂度**：TensorRT-LLM + FP8/INT4 AWQ。
- **追求灵活性和生态广度**：vLLM 或 SGLang，量化方案多、切换成本低。
- **HF 生态深度集成、要快速上线**：TGI。
- **端侧、Mac、纯 CPU、隐私部署**：llama.cpp / GGUF，没有竞品。

### 模型剪枝（Model Pruning）

剪枝通过移除模型中"不重要" 的参数、神经元或层来减小规模。

- 非结构化剪枝（Unstructured Pruning）：随机移除单个权重。虽然压缩率高，但由于硬件对稀疏矩阵支持有限，通常难以获得实际加速。
- 结构化剪枝（Structured Pruning）：移除整个通道、注意力头或网络层。这种方法能直接带来显存下降和硬件层面的推理加速。
  - 适用场景：针对特定任务（如移动端实时视觉识别）的定制化模型压缩。

### 知识蒸馏（Knowledge Distillation）

通过训练一个小型"学生模型" 来模仿大型 "教师模型" 的输出分布。适用场景：当需要一个体积极小但仍具备复杂推理能力的模型（如从 Qwen3-235B-A3B 蒸馏出高性能的 4B 模型）用于离线移动端时。

- 黑盒蒸馏：仅学习教师模型的预测标签（Logits）。
- 白盒蒸馏：学习教师模型的中间层表示（Hidden States）或注意力权重。

NVIDIA 在 2025-2026 年的研究（TensorRT Model Optimizer）强调了 Pruning -> Distillation -> Quantization（P-KD-Q）这一组合流水线是目前实现 SLM（小型语言模型）最优性能的标准路径。

而且像 LLM 教师模型的 推理能力（如工具调用）也可以教给 SLM学生模型！

### 低秩分解（Low-Rank Factorization）

将大的权重矩阵分解为两个或多个小矩阵的乘积（如 SVD 分解）。适用场景：常与 LoRA（Low-Rank Adaptation）结合，用于资源受限下的参数高效微调。

## 模型压缩工具

1. Hugging Face 生态（目前最普及的入口，通过高度抽象的接口集成了多种压缩后端）
 - Transformers & bitsandbytes：提供了"开箱即用"的量化支持，特别是 QLoRA(NF4)。它是目前微调大模型的事实标准。通过 `load_in_4bit=True` 即可调用 bitsandbytes 后端。
 - Hugging Face Optimum(v2)：2026 年推出的 Optimum v2 采用了模块化设计，将模型导出与特定硬件优化（ONNX Runtime，OpenVINO，TensorRT）解耦。其支持 Sensitivity Profiles，能自动分析模型各层对量化的敏感度，实现混合精度压缩！

2. 生产级推理加速类：NVIDIA 与 Microsoft（这类工具库追求极致的吞吐量，适用于大规模云端部署）
 - NVIDIA TensorRT-LLM & Model Optimizer：目前 NVIDIA GPU 上的性能标杆，Model Optimizer（ModelOpt）实现了 P-KD-Q（剪枝-蒸馏-量化）自动化流水线。支持最新的 FP8 量化架构（Hopper/Blackwell 显卡原生支持）以及 AutoDeploy （一键从 PyTorch 模型生成优化引擎）。
 - DeepSpeed（DeepSpeed-Compression）：微软开发的工具，强项在于 大规模稀疏化（sparsification）和权重平衡量化。其 DeepSpeed-MII 专注于低延迟推理。特别适合处理千亿级甚至万亿级参数的 MoE 模型压缩。

3. 专项量化算法类：如果需要针对特定算法进行高精度的离线量化，这些工具是首选。
 - AutoGPTQ / AutoAWQ：专注于 4-bit 量化算法的实现。AWQ（Activation-aware Weight Quantization）目前在视觉和代码模型上表现优于 GPTQ。它们产生的模型文件（如 `.safetensors`）被广泛托管在 Hugging Face 社区。
 - llama.cpp（GGUF）：统治了端侧和个人私有化部署领域。支持极其细致的量化级别（Q2_K, Q4_K_M, Q8_0 等）。支持 Apple Silicon 的 GPU 转发（Metal）和普通 CPU 上的高效推理。

4. 稀疏与剪枝类：Neural Magic
 - Neural Magic（SparseML）：专注于利用 CPU 的稀疏加速特性。在 2025 加入 RedHat 后，其技术更多集成到了 vLLM。其研发的 SparseGPT 算法可以在几小时内完成 70B 模型的结构化剪枝，且损失精度极小。
